#Loading The Dataset

In [0]:

# Loading the MNIST dataset into Keras
from keras.datasets import mnist

(train_images, train_labels), (test_images, test_labels) = mnist.load_data()


Using TensorFlow backend.


11493376/11490434 [==============================] - 1s 0us/step


In [0]:
train_images.shape

(60000, 28, 28)

In [0]:
len(train_labels)

60000

In [0]:
train_labels

array([5, 0, 4, ..., 5, 6, 8], dtype=uint8)

In [0]:
test_images.shape

(10000, 28, 28)

In [0]:
len(test_labels)

10000

In [0]:
test_labels

array([7, 2, 1, ..., 4, 5, 6], dtype=uint8)

#The Network Architecture

The core building block of a neural network is the ***layer***, a data-processing module that you can think of as a filter for data.

In the code below, our network consists of a sequence of two Dense layers, which are densely connected neural layers.

The second (to last) layer is a 10-way softmax layer as the value is 10 and activation is ***softmax***. This will return an array of 10 probability scores all summing to 1. Each will be the probability that the current digit image belongs to one of our 10 digit classes. 

In [0]:
from keras import models
from keras import layers

network = models.Sequential()
network.add(layers.Dense(512, activation='relu', input_shape=(28 * 28,)))
network.add(layers.Dense(10, activation='softmax'))

W0827 01:32:41.740226 140423971198848 deprecation_wrapper.py:119] From /usr/local/lib/python3.6/dist-packages/keras/backend/tensorflow_backend.py:74: The name tf.get_default_graph is deprecated. Please use tf.compat.v1.get_default_graph instead.

W0827 01:32:41.817749 140423971198848 deprecation_wrapper.py:119] From /usr/local/lib/python3.6/dist-packages/keras/backend/tensorflow_backend.py:517: The name tf.placeholder is deprecated. Please use tf.compat.v1.placeholder instead.

W0827 01:32:41.833202 140423971198848 deprecation_wrapper.py:119] From /usr/local/lib/python3.6/dist-packages/keras/backend/tensorflow_backend.py:4138: The name tf.random_uniform is deprecated. Please use tf.random.uniform instead.



#The Compilation Step

We need three more things for this neural network in order to begin our training:
- A *loss function* - which is how the network will measure performance on the training data and how it will be able to steer itself in the right direction.
- *Optimizer* - A mechanism which the network will update itself based on the loss function and data it sees.
- *Metrics to monitor during training and testing* - accuracy. 

In [0]:
network.compile(optimizer='rmsprop',
               loss='categorical_crossentropy',
               metrics=['accuracy'])

W0827 01:41:34.339963 140423971198848 deprecation_wrapper.py:119] From /usr/local/lib/python3.6/dist-packages/keras/optimizers.py:790: The name tf.train.Optimizer is deprecated. Please use tf.compat.v1.train.Optimizer instead.

W0827 01:41:34.371847 140423971198848 deprecation_wrapper.py:119] From /usr/local/lib/python3.6/dist-packages/keras/backend/tensorflow_backend.py:3295: The name tf.log is deprecated. Please use tf.math.log instead.



In [0]:
# Preparing the image data

train_images = train_images.reshape((60000, 28 * 28))
train_images = train_images.astype('float32') / 255

test_images = test_images.reshape((10000, 28 * 28))
test_images = test_images.astype('float32') / 255

In [0]:
# Categorically encode the labels

from keras.utils import to_categorical

train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)

#Fit The Model

In [0]:
# Ready to train the network


# Training Data Accuracy
network.fit(train_images, train_labels, epochs=5, batch_size=128)

W0827 01:45:54.918740 140423971198848 deprecation.py:323] From /usr/local/lib/python3.6/dist-packages/tensorflow/python/ops/math_grad.py:1250: add_dispatch_support.<locals>.wrapper (from tensorflow.python.ops.array_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where
W0827 01:45:54.981603 140423971198848 deprecation_wrapper.py:119] From /usr/local/lib/python3.6/dist-packages/keras/backend/tensorflow_backend.py:986: The name tf.assign_add is deprecated. Please use tf.compat.v1.assign_add instead.



Epoch 1/5
60000/60000 [==============================] - 6s 95us/step - loss: 0.2603 - acc: 0.9252
Epoch 2/5
60000/60000 [==============================] - 5s 80us/step - loss: 0.1055 - acc: 0.9689
Epoch 3/5
60000/60000 [==============================] - 5s 81us/step - loss: 0.0699 - acc: 0.9794
Epoch 4/5
60000/60000 [==============================] - 5s 81us/step - loss: 0.0510 - acc: 0.9847
Epoch 5/5
60000/60000 [==============================] - 5s 81us/step - loss: 0.0377 - acc: 0.9887


Our *training data* reached a **98.9%**. 

In [0]:
# Test Set Accuracy

test_loss, test_acc = network.evaluate(test_images, test_labels)
print('test_acc:', test_acc)

10000/10000 [==============================] - 1s 59us/step
test_acc: 0.982


Our *test data* reached a **98.2%**. 

As you notice the slight difference between accuracy scores of the train data and test data, this indicates there was a little bit of overfitting. 